File: Homework2/coin.py

def flipStack(n, coins):
    count = len(coins) - n
    return coins[:count] + list(reversed(map(lambda x: not x, coins[count:])))

def flipRec(coins):
    if all(coins):
        return [0]
    else:
        count = 1 if coins[-1] else len(coins) - coins.index(False)
        return [count] + flipRec(flipStack(count, coins))

def flipCoin(coins):
    return flipRec(list(reversed(map(lambda x: x=='H', coins))))

File: Homework2/coin.hs
import Data.List

flipStack :: Int -> [Bool] -> [Bool]
flipStack n xs = count `take` xs ++ reverse ((\x -> not x) `map` (drop count xs))
                    where count = length xs - n

flipRec :: [Bool] -> [Int]
flipRec coins
    | all (==True) coins = [0]
    | otherwise          = count : flipRec (count `flipStack` coins)
                            where count = if last coins
                                    then 1 
                                    else case (False `elemIndex` coins) of
                                        Just n -> length coins - n
                                        Nothing -> 0

flipCoin :: [Char] -> [Int]
flipCoin coins
    | all (==True) ((\x -> x=='H' || x=='T') `map` coins) = flipRec (reverse ((\x -> x == 'H') `map` coins))
    | otherwise                                           = error "Input must be H or T"
